pkg/middleware/csrf_test.go

```diff
@@ -0,0 +1,114 @@
+package middleware
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestMiddlewareCSRF(t *testing.T) {
+	tests := []struct {
+		name        string
+		cookieName  string
+		method      string
+		origin      string
+		host        string
+		defaultPort string
+		code        int
+	}{
+		{
+			name:        "mismatched origin and host is forbidden",
+			cookieName:  "foo",
+			method:      "GET",
+			origin:      "http://notLocalhost",
+			host:        "localhost",
+			defaultPort: "80",
+			code:        http.StatusForbidden,
+		},
+		{
+			name:        "mismatched origin and host is NOT forbidden with a 'Safe Method'",
+			cookieName:  "foo",
+			method:      "TRACE",
+			origin:      "http://notLocalhost",
+			host:        "localhost",
+			defaultPort: "80",
+			code:        http.StatusOK,
+		},
+		{
+			name:        "mismatched origin and host is NOT forbidden without a cookie",
+			cookieName:  "",
+			method:      "GET",
+			origin:      "http://notLocalhost",
+			host:        "localhost",
+			defaultPort: "80",
+			code:        http.StatusOK,
+		},
+		{
+			name:        "malformed host is a bad request",
+			cookieName:  "foo",
+			method:      "GET",
+			host:        "localhost:80:80",
+			defaultPort: "80",
+			code:        http.StatusBadRequest,
+		},
+		{
+			name:        "should work without port",
+			cookieName:  "foo",
+			method:      "GET",
+			host:        "localhost",
+			origin:      "http://localhost",
+			defaultPort: "80",
+			code:        http.StatusOK,
+		},
+		{
+			name:       "IPv6 host work with port",
+			cookieName: "foo",
+			method:     "GET",
+			host:       "[::1]:3000",
+			origin:     "http://[::1]:3000",
+			code:       http.StatusOK,
+		},
+		{
+			name:        "IPv6 host should get default port",
+			cookieName:  "foo",
+			method:      "GET",
+			host:        "[::1]",
+			origin:      "http://[::1]",
+			defaultPort: "3000",
+			code:        http.StatusOK,
+		},
```

Comment:
Would be nice to add the example IPv6 addresses/port from the [issue](https://github.com/grafana/grafana/issues/45115#issue-1127873033) originally reported to make sure it works for those as well.